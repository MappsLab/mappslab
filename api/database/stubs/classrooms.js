// @flow

export const aDays = {
	uid: '0x5000',
	title: "Mr. Thomas' Classroom",
}

export const edges = {}

export default []
