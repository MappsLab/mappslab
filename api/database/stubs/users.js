// @flow
export const joseph = {
	// uid: '0x1000',
	name: 'Joseph Thomas',
	role: 'student',
	email: 'joseph@good-idea.studio',
	disabled: false,
}

export const waverly = {
	// uid: '0x1001',
	name: 'Alex Johnstone',
	role: 'student',
	disabled: false,
}

export const john = {
	// uid: '0x1002',
	name: 'Mr. Schaefer',
	role: 'teacher',
	email: 'john@cmwworld.com',
}

const defaultUsers = [joseph, waverly, john]

export default defaultUsers
