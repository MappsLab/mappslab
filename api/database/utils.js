// @flow

import * as R from 'ramda'
import type { Filter, PaginationArgs } from 'Types/sharedTypes'

export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

export const defaultPaginationArgs = {
	first: 50,
	after: '0x0',
}

export const makePaginationArgs = (args: PaginationArgs): PaginationArgs => {
	// Adds +1 to the 'first' argument, we need this to see if there is a next page.
	const { first, ...rest } = { ...defaultPaginationArgs, ...args }
	return {
		first: first + 1,
		...rest,
	}
}

export const createFilterString = (filter: Array<Filter>, connect: string = 'AND'): string => {
	const funcs = filter.map(({ key, value, operator = 'eq' }) => `${operator}(${key}, "${value}")`)
	return `@filter(${funcs.join(connect)})`
}

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
