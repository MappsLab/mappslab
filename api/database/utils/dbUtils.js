// @flow

import * as R from 'ramda'
import type { Filter, PaginationFilterArgs } from 'Types/sharedTypes'
import makeFilterString from './makeFilterString'

/**
 * Properly formats incoming variables with '$' for dgraph
 */

export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

/**
 * Create a paginatino string to append to a dGraph func. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */
export const DEFAULT_PAGE_COUNT = 50

const makePaginationString = (first?: number = DEFAULT_PAGE_COUNT, after?: string): string => {
	return [
		'', // include a blank string so we start it with a comma
		first // if a 'first' value is present, include it +1 so we can tell if there is a 'nextPage'
			? `first: ${first + 1}`
			: null,
		after //
			? `after: ${after}`
			: null,
	]
		.filter((i) => i !== null)
		.join(', ')
}

/**
 * Create a @filter string to append to a dGraph query. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */

export type FilterStrings = {
	filterString: string,
	varBlocks: string,
}

type QueryStrings = FilterStrings & {
	paginationString: string,
}

export const createQueryStrings = (args: PaginationFilterArgs, filterDeleted?: boolean = true): QueryStrings => {
	if (!args)
		return {
			paginationString: '',
			filterString: '@filter((eq(deleted, false)))',
			varBlocks: '',
		}
	const { first, after, where: baseWhere } = args
	const where = filterDeleted
		? {
				deleted: { eq: false },
				...baseWhere,
		  }
		: baseWhere

	return {
		paginationString: makePaginationString(first, after),
		...makeFilterString(where),
	}
}

/**
 * Validate that the string is a dGraph uid
 */

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
