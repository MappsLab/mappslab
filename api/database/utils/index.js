export { default as makeFilterString } from './makeFilterString'
export * from './dbUtils'
