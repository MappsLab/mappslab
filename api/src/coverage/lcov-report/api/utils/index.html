<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for api/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> api/utils
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.21% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>48/57</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.85% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>28/33</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>18/22</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">85.11% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>40/47</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="auth.js"><a href="auth.js.html">auth.js</a></td>
	<td data-value="85" class="pic high"><div class="chart"><div class="cover-fill" style="width: 85%;"></div><div class="cover-empty" style="width:15%;"></div></div></td>
	<td data-value="85" class="pct high">85%</td>
	<td data-value="20" class="abs high">17/20</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	</tr>

<tr>
	<td class="file high" data-value="graphql.js"><a href="graphql.js.html">graphql.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="81.25" class="pic high"><div class="chart"><div class="cover-fill" style="width: 81%;"></div><div class="cover-empty" style="width:19%;"></div></div></td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="70" class="pct medium">70%</td>
	<td data-value="10" class="abs medium">7/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	</tr>

<tr>
	<td class="file medium" data-value="parsing.js"><a href="parsing.js.html">parsing.js</a></td>
	<td data-value="72.73" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 72%;"></div><div class="cover-empty" style="width:28%;"></div></div></td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Nov 12 2018 20:15:34 GMT-0800 (Pacific Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
