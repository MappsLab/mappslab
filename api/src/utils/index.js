// @flow
import { pipeP, flatten, filter, assoc, prop } from 'ramda'
import type { NestedArray } from 'ramda'

export const filterNullAndUndefined = (obj: Object): Promise<Object> =>
	new Promise((resolve) =>
		resolve(filter((a) => a !== null && a !== undefined)(obj)),
	)

// Makes regular functions async for promisePipe
export const promisify = (f: Function): Function => async (...args) =>
	f(...args)

export const promisePipe = (...funcs: Array<Function>) =>
	pipeP(...funcs.map(promisify))

export const arrayify = (...things: NestedArray<mixed>): Array<mixed> =>
	flatten(things)

export const slugify = (text: string) =>
	text
		.toString()
		.toLowerCase()
		.replace(/\s+/g, '-')
		.replace(/[^\w-]+/g, '')
		.replace(/--+/g, '-')
		.replace(/^-+/, '')
		.replace(/-+$/, '')

export const createSlugFrom = (key: string) => (obj: Object) =>
	assoc('slug', slugify(prop(key, obj)))(obj)
