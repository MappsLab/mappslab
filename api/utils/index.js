// @flow
import * as R from 'ramda'

export const filterNullAndUndefined = (obj: Object): Promise<Object> =>
	new Promise((resolve) => resolve(R.filter((a) => a !== null && a !== undefined)(obj)))

// Makes regular functions async for promisePipe
export const promisify = (f: Function): Function => async (...args) => f(...args)

export const promisePipe = (...funcs: Array<Function>) => R.pipeP(...funcs.map(promisify))
