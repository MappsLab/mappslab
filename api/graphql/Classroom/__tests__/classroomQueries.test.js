// @flow
/* eslint-disable no-undef */
import { artClass } from 'Database/stubs/classrooms'
import { request } from '../../../__tests__/utils'
import { getFirstClassrooms } from './utils'

let classrooms

beforeAll(async (done) => {
	classrooms = await getFirstClassrooms()
	done()
})

describe('queries', () => {
	it('[classroom] should fetch a classrom by uid', async () => {
		const query = /* GraphQL */ `
			query Classroom($uid: String) {
				classroom(input: { uid: $uid }) {
					uid
					title
				}
			}
		`
		const variables = { uid: classrooms[0].uid }
		const result = await request(query, { variables }).catch((e) => console.log(e))
		expect(result.data.classroom.title).toBe(artClass.title)
	})

	it('[classroom] should fetch a classrom by slug', async () => {
		const query = /* GraphQL */ `
			query Classroom($slug: String) {
				classroom(input: { slug: $slug }) {
					uid
					title
				}
			}
		`
		const variables = { slug: artClass.slug }
		const result = await request(query, { variables })
		expect(result.data.classroom.title).toBe(artClass.title)
	})

	it('[classroom] should fetch its connections', async () => {
		const query = /* GraphQL */ `
			query Classroom($slug: String) {
				classroom(input: { slug: $slug }) {
					uid
					title
					students {
						edges {
							node {
								name
							}
						}
					}
					teachers {
						edges {
							node {
								name
							}
						}
					}
					maps {
						edges {
							node {
								title
							}
						}
					}
				}
			}
		`
		const variables = { slug: artClass.slug }
		const result = await request(query, { variables })
		const { students, teachers, maps } = result.data.classroom
		expect(students.edges[0].node.name.length).toBeGreaterThan(0)
		expect(teachers.edges[0].node.name.length).toBeGreaterThan(0)
		expect(maps.edges[0].node.title.length).toBeGreaterThan(0)
	})

	it('[classrooms] should fetch a list of classrooms', async () => {
		const uidQuery = /* GraphQL */ `
			{
				classrooms {
					edges {
						node {
							uid
						}
					}
					pageInfo {
						lastCursor
					}
				}
			}
		`
		const uidResult = await request(uidQuery)
		expect(uidResult.data.classrooms.edges.length).toBeGreaterThan(2)
	})
})
