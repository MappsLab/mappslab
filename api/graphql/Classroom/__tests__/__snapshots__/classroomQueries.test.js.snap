// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries [classroom] should fetch its connections 1`] = `
Object {
  "node": Object {
    "name": "Matilde Prosacco MD",
  },
}
`;

exports[`queries [classroom] should fetch its connections 2`] = `
Object {
  "node": Object {
    "name": "Claudia Mohr",
  },
}
`;

exports[`queries [classroom] should fetch its connections 3`] = `
Object {
  "node": Object {
    "title": "Tasty Plastic Pizza",
  },
}
`;
